File: app/src/main/java/ua/hospes/rtm/ui/race/RaceFragment.java
package ua.hospes.rtm.ui.race;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import javax.inject.Inject;

import dagger.android.support.AndroidSupportInjection;
import ua.hospes.rtm.R;
import ua.hospes.rtm.core.StopWatchFragment;
import ua.hospes.rtm.core.StopWatchService;
import ua.hospes.rtm.domain.preferences.PreferencesManager;
import ua.hospes.rtm.domain.race.models.RaceItem;
import ua.hospes.rtm.utils.TimeUtils;
import ua.hospes.rtm.utils.UiUtils;
import ua.hospes.undobutton.UndoButton;
import ua.hospes.undobutton.UndoButtonController;

/**
 * @author Andrew Khloponin
 */
public class RaceFragment extends StopWatchFragment implements RaceContract.View {
    private static final int REQUEST_CODE_PERMISSION = 11;
    private UndoButtonController undoController;
    private TimerListController timerListController;
    @Inject RacePresenter presenter;
    @Inject PreferencesManager preferencesManager;
    private TextView tvTime;
    private RecyclerView rv;
    private RaceAdapter adapter;
    private long currentNanoTime = 0L;


    public static Fragment newInstance() {
        return new RaceFragment();
    }

    public RaceFragment() {}


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onAttach(Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
    }

    @Override
    protected int setActionBarTitle() {
        return R.string.race_title;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_race, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        rv = UiUtils.findView(view, R.id.list);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        undoController = new UndoButtonController<RaceAdapter.MyHolder>(getContext()) {
            @Override
            public UndoButton[] provideUndos(RaceAdapter.MyHolder holder) {
                return new UndoButton[]{holder.btnNextSession};
            }

            @Override
            public boolean defaultTimeSHow() {
                return true;
            }

            @Override
            public int defaultDelay() {
                return 5;
            }
        };

        rv.setHasFixedSize(true);
        rv.setLayoutManager(new LinearLayoutManager(getContext()));
        rv.setAdapter(adapter = new RaceAdapter(getContext(), preferencesManager, undoController));

        adapter.setOnPitClickListener((item, position) -> presenter.onPit(item, currentNanoTime));
        adapter.setOnOutClickListener((item, position) -> presenter.onOut(item, currentNanoTime));
        adapter.setOnUndoClickListener((item, position) -> presenter.undoLastSession(item));
        adapter.setOnSetCarClickListener((item, position) -> presenter.showSetCarDialog(getChildFragmentManager(), item.getSession()));
        adapter.setOnSetDriverClickListener((item, position) -> presenter.showSetDriverDialog(getChildFragmentManager(), item.getSession()));
        adapter.setOnItemClickListener((item, position) -> presenter.showRaceItemDetail(getContext(), item));

        rv.addOnScrollListener(timerListController = new TimerListController());
        rv.addOnScrollListener(undoController);

        presenter.attachView(this);
    }


    //region ActionBar Menu
    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.race, menu);
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        boolean stopWatchStarted = isStopWatchStarted();
        menu.findItem(R.id.action_start).setVisible(!stopWatchStarted);
        menu.findItem(R.id.action_add_team).setVisible(!stopWatchStarted);
        menu.findItem(R.id.action_export).setVisible(preferencesManager.isExportXLSEnabled() && !stopWatchStarted);
        menu.findItem(R.id.action_reset).setVisible(!stopWatchStarted);
        menu.findItem(R.id.action_clear).setVisible(!stopWatchStarted);
        menu.findItem(R.id.action_stop).setVisible(stopWatchStarted);

        MenuItem stopWatchItem = menu.findItem(R.id.action_stopwatch);
        if (stopWatchItem != null) {
            tvTime = (TextView) MenuItemCompat.getActionView(stopWatchItem).findViewById(R.id.text);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_start:
                StopWatchService.start(getContext());
                timerListController.forceUpdate(rv);
                return true;

            case R.id.action_stop:
                StopWatchService.stop(getContext());
                return true;

            case R.id.action_add_team:
                presenter.showAddTeamDialog(getChildFragmentManager());
                return true;

            case R.id.action_export:
                // Assume thisActivity is the current activity
                int permissionCheck = ContextCompat.checkSelfPermission(getContext(), Manifest.permission.WRITE_EXTERNAL_STORAGE);
                if (permissionCheck == PackageManager.PERMISSION_GRANTED) presenter.exportXLS();
                else requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, REQUEST_CODE_PERMISSION);
                return true;

            case R.id.action_reset:
                presenter.resetRace();
                return true;

            case R.id.action_clear:
                showClearDialog();
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }
    //endregion

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode) {
            case REQUEST_CODE_PERMISSION:
                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    presenter.exportXLS();
                }
                break;

            default:
                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        timerListController.unsubscribe();
        undoController.release();
        presenter.detachView();
    }


    @Override
    public void update(List<RaceItem> items) {
        adapter.clear();
        adapter.addAll(items);
    }


    @Override
    public void onStopWatchStarted(long startTime, long nanoStartTime) {
        presenter.startRace(nanoStartTime);
    }

    @Override
    public void onStopWatchStopped(long stopTime, long nanoStopTime) {
        presenter.stopRace(nanoStopTime);
    }

    @Override
    public void onStopWatchStateChanged(int runningState) {
        getActivity().supportInvalidateOptionsMenu();
    }

    @Override
    public void onStopWatchTick(long time, long nanoTime, long currentNanoTime) {
        this.currentNanoTime = currentNanoTime;
        // We have to check tvTime on null cause it couldn't be ready yet
        if (tvTime != null) tvTime.setText(TimeUtils.format(time));
        // We have to check adapter on null cause it couldn't be ready yet
        if (timerListController != null) timerListController.updateTime(currentNanoTime);
    }


    private void showClearDialog() {
        new AlertDialog.Builder(getContext())
                .setMessage(R.string.teams_remove_all)
                .setPositiveButton(R.string.yes, (dialog, which) -> presenter.removeAll())
                .setNegativeButton(R.string.no, (dialog, which) -> {})
                .show();
    }
}

File: app/src/main/java/ua/hospes/rtm/ui/race/RaceFragment.kt
package ua.hospes.rtm.ui.race

import android.Manifest
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import by.kirich1409.viewbindingdelegate.viewBinding
import dagger.hilt.android.AndroidEntryPoint
import timber.log.Timber
import ua.hospes.rtm.R
import ua.hospes.rtm.core.StopWatchFragment
import ua.hospes.rtm.core.StopWatchService
import ua.hospes.rtm.databinding.FragmentRaceBinding
import ua.hospes.rtm.domain.cars.Car
import ua.hospes.rtm.domain.drivers.Driver
import ua.hospes.rtm.domain.preferences.PreferencesManager
import ua.hospes.rtm.domain.race.models.RaceItem
import ua.hospes.rtm.ui.race.detail.intentRaceItemDetails
import ua.hospes.rtm.utils.TimeUtils
import javax.inject.Inject

private const val REQUEST_CODE_PERMISSION = 11

@AndroidEntryPoint
class RaceFragment : StopWatchFragment(R.layout.fragment_race) {
    //    private lateinit var undoController: UndoButtonController<*>
    private lateinit var timerListController: TimerListController
    private val viewModel: RaceViewModel by viewModels()
    private val binding by viewBinding(FragmentRaceBinding::bind)
    @Inject lateinit var preferencesManager: PreferencesManager
    private var tvTime: TextView? = null
    private var currentNanoTime = 0L

    private val sessionButtonType: String by lazy { preferencesManager.sessionButtonType }
    private val adapter: RaceAdapter by lazy { RaceAdapter(requireContext(), sessionButtonType/*, undoController*/) }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun setActionBarTitle(): Int = R.string.race_title

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        //        undoController = object : UndoButtonController<RaceAdapter.MyHolderUndoNext>(requireContext()) {
        //            override fun provideUndos(holder: RaceAdapter.MyHolderUndoNext) = arrayOf(holder.btnNextSession)
        //
        //            override fun defaultTimeSHow(): Boolean = true
        //
        //            override fun defaultDelay(): Int = 15
        //        }

        binding.list.setHasFixedSize(true)
        binding.list.layoutManager = LinearLayoutManager(context)
        binding.list.adapter = adapter
        binding.list.itemAnimator = null

        adapter.onPitClickListener = { viewModel.onPit(it, currentNanoTime) }
        adapter.onOutClickListener = { viewModel.onOut(it, currentNanoTime) }
        adapter.onUndoClickListener = { viewModel.undoLastSession(it) }
        adapter.setCarClickListener = { viewModel.clickSetCar(it) }
        adapter.setDriverClickListener = { viewModel.clickSetDriver(it) }
        adapter.itemClickListener = { startActivity(requireContext().intentRaceItemDetails(it.id)) }

        timerListController = TimerListController(lifecycleScope).apply { binding.list.addOnScrollListener(this) }

        //        if ("undo".equals(sessionButtonType, ignoreCase = true))
        //            list.addOnScrollListener(undoController)

        viewModel.race.observe(viewLifecycleOwner) { onData(it) }
        viewModel.uiEvents.observe(viewLifecycleOwner) {
            when (it) {
                is RaceViewModel.UIEvent.SetCar -> openSetCarDialog(it.id, it.cars)
                is RaceViewModel.UIEvent.SetDriver -> openSetDriverDialog(it.id, it.drivers)
            }
        }
    }


    //region ActionBar Menu
    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) = inflater.inflate(R.menu.race, menu)

    override fun onPrepareOptionsMenu(menu: Menu) {
        val stopWatchStarted = isStopWatchStarted
        menu.findItem(R.id.action_start).isVisible = !stopWatchStarted
        menu.findItem(R.id.action_add_team).isVisible = !stopWatchStarted
        menu.findItem(R.id.action_export).isVisible = preferencesManager.isExportXLSEnabled && !stopWatchStarted
        menu.findItem(R.id.action_reset).isVisible = !stopWatchStarted
        menu.findItem(R.id.action_clear).isVisible = !stopWatchStarted
        menu.findItem(R.id.action_stop).isVisible = stopWatchStarted

        tvTime = menu.findItem(R.id.action_stopwatch)?.actionView?.findViewById(R.id.text)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        R.id.action_start -> {
            StopWatchService.start(requireContext())
            timerListController.forceUpdate(binding.list)
            true
        }

        R.id.action_stop -> StopWatchService.stop(requireContext()).let { true }
        R.id.action_add_team -> AddTeamToRaceDialogFragment().show(childFragmentManager, "add_team_to_race").let { true }

        R.id.action_export -> {
            // Assume thisActivity is the current activity
            val permissionCheck = ContextCompat.checkSelfPermission(requireContext(), Manifest.permission.WRITE_EXTERNAL_STORAGE)
            if (permissionCheck == PackageManager.PERMISSION_GRANTED)
                viewModel.exportXLS()
            else
                requestPermissions(arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), REQUEST_CODE_PERMISSION)
            true
        }

        R.id.action_reset -> viewModel.resetRace().let { true }
        R.id.action_clear -> showClearDialog().let { true }
        else -> super.onOptionsItemSelected(item)
    }
    //endregion

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) = when (requestCode) {
        REQUEST_CODE_PERMISSION -> if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            viewModel.exportXLS()
        } else Unit

        else -> super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        //        undoController.release()
    }

    private fun onData(items: List<RaceItem>) {
        Timber.d(items.toString())
        adapter.submitList(items)
        timerListController.forceUpdate(binding.list)
        //        if ("undo".equals(sessionButtonType, ignoreCase = true))
        //            undoController.forceUpdate(list)
    }

    // override fun onError(t: Throwable) = Toast.makeText(context, t.message, Toast.LENGTH_SHORT).show()


    override fun onStopWatchStarted(startTime: Long, nanoStartTime: Long) = viewModel.startRace(nanoStartTime).let { Unit }
    override fun onStopWatchStopped(stopTime: Long, nanoStopTime: Long) = viewModel.stopRace(nanoStopTime).let { Unit }
    override fun onStopWatchStateChanged(runningState: Int) = activity?.invalidateOptionsMenu() ?: Unit

    override fun onStopWatchTick(time: Long, nanoTime: Long, currentNanoTime: Long) {
        this.currentNanoTime = currentNanoTime
        // We have to check tvTime on null cause it couldn't be ready yet
        tvTime?.text = TimeUtils.format(time)
        // We have to check adapter on null cause it couldn't be ready yet
        timerListController.updateTime(currentNanoTime)
    }


    private fun openSetCarDialog(sessionId: Long, cars: List<Car>) =
        SetCarDialogFragment.newInstance(sessionId, cars).show(childFragmentManager, "set_car")

    private fun openSetDriverDialog(sessionId: Long, drivers: List<Driver>) =
        SetDriverDialogFragment.newInstance(sessionId, drivers).show(childFragmentManager, "set_driver")


    private fun showClearDialog() = AlertDialog.Builder(requireContext())
        .setMessage(R.string.teams_remove_all)
        .setPositiveButton(R.string.yes) { _, _ -> viewModel.removeAll() }
        .setNegativeButton(R.string.no) { _, _ -> }
        .show()
}